buffer: remove async tick and improve perf by 48%

anonrig opened the pull request:
I ran the benchmarks only for `blob.text()` since it's the only relevant one, but added more for future reference.

```
                                                       confidence improvement accuracy (*)   (**)  (***)
blob/blob-text.js operation='text' n=1000000 bytes=0          ***     48.33 %       ±1.65% ±2.20% ±2.87%
blob/blob-text.js operation='text' n=1000000 bytes=128        ***      9.65 %       ±2.33% ±3.10% ±4.04%
blob/blob-text.js operation='text' n=1000000 bytes=8          ***      8.20 %       ±1.84% ±2.45% ±3.19%

Be aware that when doing many comparisons the risk of a false-positive result increases.
In this case, there are 3 comparisons, you can thus expect the following amount of false-positive results:
  0.15 false positives, when considering a   5% risk acceptance (*, **, ***),
  0.03 false positives, when considering a   1% risk acceptance (**, ***),
  0.00 false positives, when considering a 0.1% risk acceptance (***)
```

RafaelGSS commented:
@anonrig Did you run the benchmark against `main` or does it applies to other release lines too?

anonrig commented:
@RafaelGSS I run the command only on `main` using the following command: `node benchmark/compare.js --old ./node-master --new ./out/Release/node blob > blob.csv`. IMHO, I'm pretty sure it applies to all release lines.

addaleax commented on benchmark/blob/blob-text.js:
```diff
@@ -0,0 +1,29 @@
+        break;
+    }
+  }
+  bench.end(n);
```
`fs/bench-stat-promise.js` might be a better candidate – but generally, just making sure that we're actually `await`ing the results here and including that in the benchmark time, i.e. making this function `async` and adding `await` where necessary, should be enough.

anonrig replied:
I force pushed and updated the benchmarks. Thanks for the explanation.

anonrig replied:
I'll update the benchmark results soon.

addaleax commented on benchmark/blob/blob-text.js:
```diff
@@ -0,0 +1,33 @@
+        await source.arrayBuffer();
+        break;
+      case 'stream':
+        source.stream();
```
Keep in mind that this doesn't properly compare to the other operations without actually consuming the data – but I'd leave it up to you whether you want to include that or not. Without it, this part of benchmark isn't very meaningful, it measures stream creation time rather than data reading time, but it does compare meaningfully across versions in that way.

anonrig replied:
Yes, promisifying this is the correct approach. I'll update this in a new pull request, and for now just focus on opening similar pull requests and reducing the ticks.

aduh95 commented:
Benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/1195/

anonrig commented:
I updated the benchmark result and the title according to the changes in the benchmark.

BridgeAR commented on benchmark/blob/blob-text.js:
```diff
@@ -0,0 +1,33 @@
+const { Blob } = require('buffer');
+
+const bench = common.createBenchmark(main, {
+  bytes: [0, 8, 128],
```
128 bytes seems rather small. I guess this is mainly used for bigger values such as 128kb, 1mb and similar.

BridgeAR reviewed (commented):
I am uncertain if it's really worth it as it will likely only improve the situation for very small entries that won't be the common case.
If it's really an API that is called millions of times with very small values, it's a different story. Using promise .then() has the downside of not providing async stack traces. Thus, I am -0.5.

mcollina reviewed (approved):
lgtm

aduh95 commented:
The benchmark CI didn't show any perf improvement with this PR 🤔

anonrig commented:
My branch was not up to date (by 35 commits). I rebased, built main & branch from scratch, force-pushed to this branch, removed `stream` from the benchmark, and re-run the benchmark on my machine. The results are similar. I don't have logical reasoning behind this difference. Can you re-run the benchmark CI @aduh95? If the results are the same, I'll just close the pull request.

anonrig commented:
Can we run the benchmark on v18 and v16 branches instead of main? I suspect v8 version bump fixed this regression.

aduh95 commented:
Benchmark CI (main): https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/1196/
Benchmark CI (v18.x): https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/1197/
Benchmark CI (v16.x): https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/1198/

_N.B.: some of the above CIs are queued, the links will 404 until it starts._

BridgeAR reviewed (commented):
The benchmark shows no improvement:

```
23:50:37                                                               confidence improvement accuracy (*)   (**)   (***)
23:50:38 blob/blob-text.js operation='arrayBuffer' n=1000000 bytes=0                   1.80 %       ±4.26% ±5.68%  ±7.40%
23:50:38 blob/blob-text.js operation='arrayBuffer' n=1000000 bytes=128                -0.31 %       ±4.68% ±6.22%  ±8.10%
23:50:38 blob/blob-text.js operation='arrayBuffer' n=1000000 bytes=8                  -2.30 %       ±5.96% ±7.93% ±10.32%
23:50:38 blob/blob-text.js operation='text' n=1000000 bytes=0                         -3.52 %       ±5.14% ±6.83%  ±8.90%
23:50:38 blob/blob-text.js operation='text' n=1000000 bytes=128                        0.61 %       ±6.97% ±9.27% ±12.07%
23:50:38 blob/blob-text.js operation='text' n=1000000 bytes=8                         -1.03 %       ±2.79% ±3.71%  ±4.85%
```

The patch does not apply cleanly to older versions and can't automatically be checked against those versions due to that.

If this would really change anything, it would only be noticeable for super tiny blobs and they should not be of any concern. As such, I recommend to stick to what we have.

anonrig commented:
Sorry for the late reply. I rebuild this change on both v16 and v18, and I could not reproduce the same benchmark result on either branch. Therefore, I'm closing this pull request and opening a pull-request only for the benchmark: https://github.com/nodejs/node/pull/44990. 

I appreciate anyone who can help me understand how I got the initial 640% and 40% diff on benchmarks. Nevertheless, Thank you all for the reviews and comments.
